core/desugarDatatypes.ml

```diff
@@ -209,6 +213,72 @@ module Desugar = struct
     let seed =
       let open Datatype in
       match rv with
+        | EffectApplication (name, ts) ->
+            let match_quantifiers : type a. (a -> Kind.t) -> a list -> Types.type_arg list = fun proj qs ->
+              let match_kinds i (q, t) =
+                let primary_kind_of_type_arg : Datatype.type_arg -> PrimaryKind.t = function
+                  | Type _ -> PrimaryKind.Type
+                  | Row _ -> PrimaryKind.Row
+                  | Presence _ -> PrimaryKind.Presence
+                in
+                let q_kind, _ = proj q in
+                let t_kind = primary_kind_of_type_arg t in
+                if q_kind <> t_kind then
+                  raise
+                    (TypeApplicationKindMismatch
+                       { pos = node.pos;
+                         name = name;
+                         tyarg_number = i;
+                         expected = PrimaryKind.to_string q_kind;
+                         provided = PrimaryKind.to_string t_kind
+                       })
+                else t
+              in
+              let type_args qs ts =
+                List.combine qs ts
+                |> List.mapi
+                     (fun i (q,t) ->
+                       let  t = match_kinds i (q, t) in
+                       type_arg alias_env t node)
+              in
+
+              let qn = List.length qs and tn = List.length ts in
+              if qn = tn then
+                type_args qs ts
+              else
+                raise (TypeApplicationArityMismatch { pos = node.pos; name = name; expected = qn; provided = tn })
+            in
+            begin match SEnv.find_opt name alias_env with
+              | None -> raise (UnboundTyCon (node.pos, name))
```

Comment:
`UnboundTyCon` should probably also be defined in `errors.ml`

Reply:
idem